### ISSUE TITLE
DefaultDiagnosticsContext should cache resultSetFetchedRows

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14191]
The logic in `DefaultDiagnosticsContext::resultSetFetchedRows` alters the underlying `ResultSet` in order to calculate the actual fetched rows:

```java
    public final int resultSetFetchedRows() {
        if (resultSet == null)
            return -1;

        try {
            if (resultSetClosing || resultSet.getType() != ResultSet.TYPE_FORWARD_ONLY) {
                while (resultSet.next())
                    resultSetFetchedRows++;

                resultSet.absolute(resultSetConsumedRows);
            }
        }
        catch (SQLException ignore) {}

        return resultSetFetchedRows;
    }
```

This means that one `DiagnosticsListener` querying this value might affect the result of this method for other listeners later on.

**Comment by lukaseder** [comment]
This was discovered as the `LoggingDiagnosticsListener` created a regression: https://github.com/jOOQ/jOOQ/issues/14146

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.6 (https://github.com/jOOQ/jOOQ/issues/14192), and 3.16.12 (https://github.com/jOOQ/jOOQ/issues/14193)
